coc/abc.py

```diff
@@ -226,75 +228,79 @@ def _load_json_meta(cls, json_meta, id, name, lab_to_townhall):
                 laboratory_levels = []
-                for lab_level in json_meta.get("LaboratoryLevel"):
+                for lab_level in [json_meta.get(level).get("LaboratoryLevel") for level in levels_available]:
                     laboratory_levels.append(max(lab_level, first_lab_level))
+
             else:
                 return
 
         cls.lab_level = try_enum(UnitStat, laboratory_levels)
-        cls.housing_space = _get_maybe_first(json_meta, "HousingSpace", default=0)
-        cls.speed = try_enum(UnitStat, json_meta.get("Speed"))
+        cls.housing_space = json_meta.get("HousingSpace", 0)
+
+        cls.speed = try_enum(UnitStat, [json_meta.get(level).get("Speed") for level in levels_available])
         cls.level = cls.dps and UnitStat(range(1, len(cls.dps) + 1))
 
-        # all 3
-        cls.upgrade_cost = try_enum(UnitStat, json_meta.get("UpgradeCost"))
-        cls.upgrade_resource = Resource(value=json_meta["UpgradeResource"][0])
-        cls.upgrade_time = try_enum(UnitStat,
-                                    [TimeDelta(hours=hours) for hours in
-                                     json_meta.get("UpgradeTimeH", [])])
+        cls.upgrade_cost = try_enum(UnitStat, [json_meta.get(level).get("UpgradeCost") for level in levels_available])
+        cls.upgrade_resource = Resource(value=json_meta.get("UpgradeResource"))
+        upgrade_times = [
+            TimeDelta(hours=hours)
+            for level in levels_available
+            if (hours := json_meta.get(level, {}).get("UpgradeTimeH")) is not None
+        ]
+        cls.upgrade_time = try_enum(UnitStat, upgrade_times)
+
         cls._is_home_village = False if json_meta.get("VillageType") else True
         cls.village = "home" if cls._is_home_village else "builderBase"
 
-        # spells and troops
-        cls.training_cost = try_enum(UnitStat, json_meta.get("TrainingCost"))
-        cls.training_time = try_enum(UnitStat, json_meta.get("TrainingTime"))
+        cls.training_time = TimeDelta(seconds=json_meta.get("TrainingTime"))
 
         # only heroes
-        cls.ability_time = try_enum(UnitStat, json_meta.get("AbilityTime"))
-        cls.ability_troop_count = try_enum(UnitStat, json_meta.get("AbilitySummonTroopCount"))
-        cls.required_th_level = try_enum(UnitStat, json_meta.get("RequiredTownHallLevel") or laboratory_levels)
-        cls.regeneration_time = try_enum(
-            UnitStat, [TimeDelta(minutes=value) for value in json_meta.get("RegenerationTimeMinutes", [])]
-        )
+        cls.ability_time = try_enum(UnitStat, [json_meta.get(level).get("AbilityTime") for level in levels_available])
+        cls.ability_troop_count = try_enum(UnitStat, [json_meta.get(level).get("AbilitySummonTroopCount") for level in levels_available])
+
+        required_townhall_levels = [json_meta.get(level).get("RequiredTownHallLevel") for level in levels_available]
```

Comment:
does this already reflect the hero hall dependency instead of the former th one?